{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/broad/blainey_lab/blainey-ipython/packages/matplotlib/__init__.py:1401: UserWarning:  This call to matplotlib.use() has no effect\n",
      "because the backend has already been chosen;\n",
      "matplotlib.use() must be called *before* pylab, matplotlib.pyplot,\n",
      "or matplotlib.backends is imported for the first time.\n",
      "\n",
      "  warnings.warn(_use_error_msg)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "# for plotting\n",
    "import seaborn as sns\n",
    "sns.set(style='white', font_scale=1.5)\n",
    "#for working with Bam files\n",
    "import pysam\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.backends.backend_pdf\n",
    "#for distance\n",
    "from scipy.spatial import distance\n",
    "import datetime\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "\n",
    "import logging\n",
    "from gridmap import grid_map\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step1:\n",
    "## Collect the results from the SNP caller\n",
    "## and modify it to get one big list of optional candidates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/broad/mushroom_lasagna/YB/Ht115_new_samples\n"
     ]
    }
   ],
   "source": [
    "cd /broad/mushroom_lasagna/YB/Ht115_new_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T34_N48.txt\n"
     ]
    }
   ],
   "source": [
    "################################################################\n",
    "#create the header file _ by using the original output file\n",
    "frames=[]\n",
    "out_files = !ls | grep 'T34_N48.txt'\n",
    "f=out_files[0]\n",
    "print f\n",
    "!grep contig {f} > /broad/mushroom_lasagna/YB/{'header2.txt'} \n",
    "################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/broad/blainey_lab/blainey-ipython/anaconda/lib/python2.7/site-packages/ipykernel_launcher.py:9: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.\n",
      "  if __name__ == '__main__':\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('for: ', 'T34_N38.txt', ' -The size is: ', (4773, 51))\n",
      "('for: ', 'T34_N44.txt', ' -The size is: ', (4422, 51))\n",
      "('for: ', 'T34_N45.txt', ' -The size is: ', (5023, 51))\n",
      "('for: ', 'T34_N47.txt', ' -The size is: ', (5093, 51))\n",
      "('for: ', 'T34_N48.txt', ' -The size is: ', (4812, 51))\n",
      "('for: ', 'T34_N49.txt', ' -The size is: ', (5029, 51))\n",
      "('for: ', 'T34_N54.txt', ' -The size is: ', (5193, 51))\n",
      "('for: ', 'T34_N56.txt', ' -The size is: ', (5154, 51))\n",
      "('for: ', 'T34_N57.txt', ' -The size is: ', (4572, 51))\n",
      "('for: ', 'T34_N63.txt', ' -The size is: ', (5130, 51))\n",
      "('for: ', 'T38_N34.txt', ' -The size is: ', (3167, 51))\n",
      "('for: ', 'T38_N44.txt', ' -The size is: ', (3352, 51))\n",
      "('for: ', 'T38_N45.txt', ' -The size is: ', (2952, 51))\n",
      "('for: ', 'T38_N47.txt', ' -The size is: ', (3475, 51))\n",
      "('for: ', 'T38_N48.txt', ' -The size is: ', (2800, 51))\n",
      "('for: ', 'T38_N49.txt', ' -The size is: ', (4001, 51))\n",
      "('for: ', 'T38_N54.txt', ' -The size is: ', (3522, 51))\n",
      "('for: ', 'T38_N56.txt', ' -The size is: ', (3055, 51))\n",
      "('for: ', 'T38_N57.txt', ' -The size is: ', (2672, 51))\n",
      "('for: ', 'T38_N63.txt', ' -The size is: ', (4027, 51))\n",
      "('for: ', 'T44_N34.txt', ' -The size is: ', (2487, 51))\n",
      "('for: ', 'T44_N38.txt', ' -The size is: ', (3145, 51))\n",
      "('for: ', 'T44_N45.txt', ' -The size is: ', (3446, 51))\n",
      "('for: ', 'T44_N47.txt', ' -The size is: ', (3510, 51))\n",
      "('for: ', 'T44_N48.txt', ' -The size is: ', (3303, 51))\n",
      "('for: ', 'T44_N49.txt', ' -The size is: ', (3499, 51))\n",
      "('for: ', 'T44_N54.txt', ' -The size is: ', (3476, 51))\n",
      "('for: ', 'T44_N56.txt', ' -The size is: ', (3325, 51))\n",
      "('for: ', 'T44_N57.txt', ' -The size is: ', (3094, 51))\n",
      "('for: ', 'T44_N63.txt', ' -The size is: ', (3558, 51))\n",
      "('for: ', 'T45_N34.txt', ' -The size is: ', (3078, 51))\n",
      "('for: ', 'T45_N38.txt', ' -The size is: ', (2680, 51))\n",
      "('for: ', 'T45_N44.txt', ' -The size is: ', (3482, 51))\n",
      "('for: ', 'T45_N47.txt', ' -The size is: ', (3414, 51))\n",
      "('for: ', 'T45_N48.txt', ' -The size is: ', (2539, 51))\n",
      "('for: ', 'T45_N49.txt', ' -The size is: ', (3852, 51))\n",
      "('for: ', 'T45_N54.txt', ' -The size is: ', (3494, 51))\n",
      "('for: ', 'T45_N56.txt', ' -The size is: ', (3150, 51))\n",
      "('for: ', 'T45_N57.txt', ' -The size is: ', (2732, 51))\n",
      "('for: ', 'T45_N63.txt', ' -The size is: ', (3985, 51))\n",
      "('for: ', 'T47_N34.txt', ' -The size is: ', (3472, 51))\n",
      "('for: ', 'T47_N38.txt', ' -The size is: ', (3598, 51))\n",
      "('for: ', 'T47_N44.txt', ' -The size is: ', (3928, 51))\n",
      "('for: ', 'T47_N45.txt', ' -The size is: ', (3835, 51))\n",
      "('for: ', 'T47_N48.txt', ' -The size is: ', (3612, 51))\n",
      "('for: ', 'T47_N49.txt', ' -The size is: ', (4192, 51))\n",
      "('for: ', 'T47_N54.txt', ' -The size is: ', (2940, 51))\n",
      "('for: ', 'T47_N56.txt', ' -The size is: ', (3897, 51))\n",
      "('for: ', 'T47_N57.txt', ' -The size is: ', (3451, 51))\n",
      "('for: ', 'T47_N63.txt', ' -The size is: ', (4325, 51))\n",
      "('for: ', 'T48_N34.txt', ' -The size is: ', (3622, 51))\n",
      "('for: ', 'T48_N38.txt', ' -The size is: ', (3265, 51))\n",
      "('for: ', 'T48_N44.txt', ' -The size is: ', (4106, 51))\n",
      "('for: ', 'T48_N45.txt', ' -The size is: ', (3377, 51))\n",
      "('for: ', 'T48_N47.txt', ' -The size is: ', (3973, 51))\n",
      "('for: ', 'T48_N49.txt', ' -The size is: ', (4423, 51))\n",
      "('for: ', 'T48_N54.txt', ' -The size is: ', (4137, 51))\n",
      "('for: ', 'T48_N56.txt', ' -The size is: ', (3789, 51))\n",
      "('for: ', 'T48_N57.txt', ' -The size is: ', (3309, 51))\n",
      "('for: ', 'T48_N63.txt', ' -The size is: ', (4598, 51))\n",
      "('for: ', 'T49_N34.txt', ' -The size is: ', (3848, 51))\n",
      "('for: ', 'T49_N38.txt', ' -The size is: ', (4458, 51))\n",
      "('for: ', 'T49_N44.txt', ' -The size is: ', (4356, 51))\n",
      "('for: ', 'T49_N45.txt', ' -The size is: ', (4592, 51))\n",
      "('for: ', 'T49_N47.txt', ' -The size is: ', (4611, 51))\n",
      "('for: ', 'T49_N48.txt', ' -The size is: ', (4456, 51))\n",
      "('for: ', 'T49_N54.txt', ' -The size is: ', (4794, 51))\n",
      "('for: ', 'T49_N56.txt', ' -The size is: ', (4931, 51))\n",
      "('for: ', 'T49_N57.txt', ' -The size is: ', (4254, 51))\n",
      "('for: ', 'T49_N63.txt', ' -The size is: ', (3880, 51))\n",
      "('for: ', 'T54_N34.txt', ' -The size is: ', (3169, 51))\n",
      "('for: ', 'T54_N38.txt', ' -The size is: ', (3212, 51))\n",
      "('for: ', 'T54_N44.txt', ' -The size is: ', (3541, 51))\n",
      "('for: ', 'T54_N45.txt', ' -The size is: ', (3469, 51))\n",
      "('for: ', 'T54_N47.txt', ' -The size is: ', (2536, 51))\n",
      "('for: ', 'T54_N48.txt', ' -The size is: ', (3314, 51))\n",
      "('for: ', 'T54_N49.txt', ' -The size is: ', (3939, 51))\n",
      "('for: ', 'T54_N56.txt', ' -The size is: ', (3531, 51))\n",
      "('for: ', 'T54_N57.txt', ' -The size is: ', (3137, 51))\n",
      "('for: ', 'T54_N63.txt', ' -The size is: ', (4002, 51))\n",
      "('for: ', 'T56_N34.txt', ' -The size is: ', (3021, 51))\n",
      "('for: ', 'T56_N38.txt', ' -The size is: ', (2711, 51))\n",
      "('for: ', 'T56_N44.txt', ' -The size is: ', (3233, 51))\n",
      "('for: ', 'T56_N45.txt', ' -The size is: ', (2891, 51))\n",
      "('for: ', 'T56_N47.txt', ' -The size is: ', (3310, 51))\n",
      "('for: ', 'T56_N48.txt', ' -The size is: ', (2834, 51))\n",
      "('for: ', 'T56_N49.txt', ' -The size is: ', (3831, 51))\n",
      "('for: ', 'T56_N54.txt', ' -The size is: ', (3277, 51))\n",
      "('for: ', 'T56_N57.txt', ' -The size is: ', (2415, 51))\n",
      "('for: ', 'T56_N63.txt', ' -The size is: ', (3825, 51))\n",
      "('for: ', 'T57_N34.txt', ' -The size is: ', (3734, 51))\n",
      "('for: ', 'T57_N38.txt', ' -The size is: ', (3526, 51))\n",
      "('for: ', 'T57_N44.txt', ' -The size is: ', (4117, 51))\n",
      "('for: ', 'T57_N45.txt', ' -The size is: ', (3882, 51))\n",
      "('for: ', 'T57_N47.txt', ' -The size is: ', (4169, 51))\n",
      "('for: ', 'T57_N48.txt', ' -The size is: ', (3677, 51))\n",
      "('for: ', 'T57_N49.txt', ' -The size is: ', (4608, 51))\n",
      "('for: ', 'T57_N54.txt', ' -The size is: ', (4283, 51))\n",
      "('for: ', 'T57_N56.txt', ' -The size is: ', (3705, 51))\n",
      "('for: ', 'T57_N63.txt', ' -The size is: ', (4706, 51))\n",
      "('for: ', 'T63_N34.txt', ' -The size is: ', (2824, 51))\n",
      "('for: ', 'T63_N38.txt', ' -The size is: ', (3354, 51))\n",
      "('for: ', 'T63_N44.txt', ' -The size is: ', (3166, 51))\n",
      "('for: ', 'T63_N45.txt', ' -The size is: ', (3569, 51))\n",
      "('for: ', 'T63_N47.txt', ' -The size is: ', (3500, 51))\n",
      "('for: ', 'T63_N48.txt', ' -The size is: ', (3396, 51))\n",
      "('for: ', 'T63_N49.txt', ' -The size is: ', (2788, 51))\n",
      "('for: ', 'T63_N54.txt', ' -The size is: ', (3686, 51))\n",
      "('for: ', 'T63_N56.txt', ' -The size is: ', (3648, 51))\n",
      "('for: ', 'T63_N57.txt', ' -The size is: ', (3226, 51))\n",
      "(406264, 51)\n"
     ]
    }
   ],
   "source": [
    "#load all files into one big table - sort by \n",
    "#not_contain_the outer sample\n",
    "frames=[]\n",
    "out_files = !ls | grep 'T' | grep '.txt' | grep -v '37'\n",
    "count=1\n",
    "columns = pd.read_csv('/broad/mushroom_lasagna/YB/header.txt', sep='\\t', comment='#').columns\n",
    "for file in out_files:\n",
    "    wt = pd.read_csv(file, sep='\\t')\n",
    "    wt = pd.DataFrame(wt.as_matrix(), columns=columns)\n",
    "    print (\"for: \",file,\" -The size is: \",wt.shape)\n",
    "    if (count==1):\n",
    "        wf=wt\n",
    "    else:\n",
    "        wf=wf.append(wt, ignore_index=True)\n",
    "    count=count+1\n",
    "print (wf.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cwf=wf.drop_duplicates(['position','contig'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full size of the list: 64770\n"
     ]
    }
   ],
   "source": [
    "f1=open('/broad/mushroom_lasagna/YB/Ht115_new_samples/log.txt', 'w+')\n",
    "co=1\n",
    "columns=('contig','position','context','ref_all','alt_all','t_lod_fstar','vector','distance_in_g1','distance_in_g2','per_bet_gr','real_bet_gr','ref_af','alt_af','type','full_data')\n",
    "mutation=[]\n",
    "f1.write(\"Log for run: /n\")\n",
    "keep=['contig', 'position','context', 'ref_allele','alt_allele','t_lod_fstar']\n",
    "locations=cwf[keep]\n",
    "print \"full size of the list:\",locations.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the names for all bams file for every sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>individual_id</th>\n",
       "      <th>clean_bam_file_capture</th>\n",
       "      <th>short_letter_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T37</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/seq/picard_aggregation/G101103/HT6W37/v4/HT6W...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T45</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/seq/picard_aggregation/G101103/HT6W45/v4/HT6W...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T47</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/seq/picard_aggregation/G101103/HT6W47/v4/HT6W...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T48</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/seq/picard_aggregation/G101103/HT6W48/v4/HT6W...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T49</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/seq/picard_aggregation/G101103/HT6W49/v4/HT6W...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T54</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/seq/picard_aggregation/G101103/HT6W54/v4/HT6W...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T34</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/seq/picard_aggregation/G101103/HT6W34/v4/HT6W...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T63</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/seq/picard_aggregation/G101103/HT6W63/v4/HT6W...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T56</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/broad/blainey_lab_storage/YB/G101103/Second_B...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T38</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/broad/blainey_lab_storage/YB/G101103/Second_B...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>T44</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/broad/blainey_lab_storage/YB/G101103/Second_B...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>T57</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/broad/blainey_lab_storage/YB/G101103/Second_B...</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CCLE</td>\n",
       "      <td>HT115_6</td>\n",
       "      <td>/seq/picard_aggregation/G32831/HT115/v2/HT115.bam</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id        ...        short_letter_code\n",
       "0        T37        ...                        T\n",
       "1        T45        ...                        T\n",
       "2        T47        ...                        T\n",
       "3        T48        ...                        T\n",
       "4        T49        ...                        T\n",
       "5        T54        ...                        T\n",
       "6        T34        ...                        T\n",
       "7        T63        ...                        T\n",
       "8        T56        ...                        T\n",
       "9        T38        ...                        T\n",
       "10       T44        ...                        T\n",
       "11       T57        ...                        T\n",
       "12      CCLE        ...                        T\n",
       "\n",
       "[13 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = pd.read_csv('/broad/mushroom_lasagna/YB/Ht115_new_samples/new_HT_samples.csv')\n",
    "print files.shape\n",
    "files[0:13]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For every SNV, collect all reads from the samples (bam files). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#\n",
    "#and save the table - should run only once.\n",
    "#def mat_from_bam(locations):\n",
    "# this have to be in seperate line - ??????\n",
    "#!cd /seq/picard_aggregation/G101103/\n",
    "f1=open('/broad/mushroom_lasagna/YB/RPE_new_samples/log.txt', 'w+')\n",
    "f1.write(\"Log for run: /n\")\n",
    "\n",
    "head=[]\n",
    "begin=!pwd \n",
    "begin=begin.pop()\n",
    "\n",
    "#prepare list of all bam files to read\n",
    "#open bam file and load all the locations - and add to the table\n",
    "for ind in files.index:\n",
    "    head, tail =os.path.split(files.clean_bam_file_capture[ind])\n",
    "    print head\n",
    "    os.chdir(head)\n",
    "    path=files.clean_bam_file_capture[ind]\n",
    "    samfile = pysam.AlignmentFile(path, \"rb\")\n",
    "    #Samples.append(samfile)\n",
    "    #for every bam file run over all the mutations\n",
    "    seq=[]\n",
    "    f1.write(\"Reading from file: %s\\n\"%(tail))\n",
    "    f1.flush()\n",
    "    for i in locations.index:\n",
    "        if (i%500 ==0):\n",
    "            f1.write(\"file %s finished = %i lines\\n\" %(tail,i))\n",
    "            f1.flush()\n",
    "        s=\"\"\n",
    "        chr= str(locations.contig[i])\n",
    "        region= int(float(locations.position[i]))\n",
    "        region=region-1\n",
    "        for row in samfile.pileup(chr,region,region+1):\n",
    "                if row.pos ==region:\n",
    "                    for row in row.pileups:\n",
    "                        if (row.query_position is not None):\n",
    "                            #print (row.query_position)\n",
    "                            #print (row.alignment.query_sequence[row.query_position])\n",
    "                            s+=row.alignment.query_sequence[row.query_position]\n",
    "        s=''.join(sorted(s.upper()))\n",
    "        seq.append(s)\n",
    "        # add the seq as column in the db - location\n",
    "    #locations[f]=pd.Series(seq,index=f)\n",
    "    ntail = tail.split('.')[0]\n",
    "    #print \"ntail= \",ntail,\" len seq= \",len(seq)\n",
    "    locations[ntail]=seq\n",
    "\n",
    "########################\n",
    "#and save the new table\n",
    "locations.to_pickle('/broad/mushroom_lasagna/YB/HT11_new_samples/mutations.pkl')\n",
    "f1.write(\"file %s finished = %i lines\\n\" %(ntail,i))\n",
    "f1.flush()\n",
    "f1.write(\"finito\")\n",
    "f1.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>position</th>\n",
       "      <th>context</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>t_lod_fstar</th>\n",
       "      <th>HT6W37</th>\n",
       "      <th>HT6W45</th>\n",
       "      <th>HT6W47</th>\n",
       "      <th>HT6W48</th>\n",
       "      <th>HT6W49</th>\n",
       "      <th>HT6W54</th>\n",
       "      <th>HT6W34</th>\n",
       "      <th>HT6W63</th>\n",
       "      <th>HT115_42112</th>\n",
       "      <th>HT115_42221</th>\n",
       "      <th>HT115_5211</th>\n",
       "      <th>HT115_7221</th>\n",
       "      <th>HT115</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1131218</td>\n",
       "      <td>TCTxCCT</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>7.95299</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>ACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1531286</td>\n",
       "      <td>ACAxAGA</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>6.83966</td>\n",
       "      <td>ACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCGGGGGGGG</td>\n",
       "      <td>AACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGT</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCGGGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGGGGGGG</td>\n",
       "      <td>AACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1894948</td>\n",
       "      <td>AAAxAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>7.7051</td>\n",
       "      <td>AAAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCG</td>\n",
       "      <td>AAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCG</td>\n",
       "      <td>ACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCG</td>\n",
       "      <td>AACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAAAAAAAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>AAAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>ACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contig              ...                                        HT115\n",
       "0      1              ...                  CCCCCCCCCCCCCCCCCCCCCCCCCCC\n",
       "1      1              ...                CCCCCCCCCCCCCCCCCCCCCCCCCCCCG\n",
       "2      1              ...                         CCCCCCCCCCCCCCCCCCCC\n",
       "\n",
       "[3 rows x 19 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "locations[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(64770, 19)\n",
      "['HT6W37', 'HT6W45', 'HT6W47', 'HT6W48', 'HT6W49', 'HT6W54', 'HT6W34', 'HT6W63', 'HT6W56', 'HT6W38', 'HT6W44', 'HT6W57', 'HT_CCLE']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>position</th>\n",
       "      <th>context</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>t_lod_fstar</th>\n",
       "      <th>HT6W37</th>\n",
       "      <th>HT6W45</th>\n",
       "      <th>HT6W47</th>\n",
       "      <th>HT6W48</th>\n",
       "      <th>HT6W49</th>\n",
       "      <th>HT6W54</th>\n",
       "      <th>HT6W34</th>\n",
       "      <th>HT6W63</th>\n",
       "      <th>HT6W56</th>\n",
       "      <th>HT6W38</th>\n",
       "      <th>HT6W44</th>\n",
       "      <th>HT6W57</th>\n",
       "      <th>HT_CCLE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1131218</td>\n",
       "      <td>TCTxCCT</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>7.95299</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>ACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1531286</td>\n",
       "      <td>ACAxAGA</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>6.83966</td>\n",
       "      <td>ACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCGGGGGGGG</td>\n",
       "      <td>AACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGT</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCGGGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGGGGGGG</td>\n",
       "      <td>AACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGG</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCCCCCCCCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1894948</td>\n",
       "      <td>AAAxAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>7.7051</td>\n",
       "      <td>AAAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCG</td>\n",
       "      <td>AAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCG</td>\n",
       "      <td>ACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCG</td>\n",
       "      <td>AACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAAAAAAAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>AAAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>AAACCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>ACCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contig              ...                                      HT_CCLE\n",
       "0      1              ...                  CCCCCCCCCCCCCCCCCCCCCCCCCCC\n",
       "1      1              ...                CCCCCCCCCCCCCCCCCCCCCCCCCCCCG\n",
       "2      1              ...                         CCCCCCCCCCCCCCCCCCCC\n",
       "\n",
       "[3 rows x 19 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#For HT115:\n",
    "#rename all the samples you are going to analyse\n",
    "\n",
    "print locations.shape\n",
    "#locations =locations.drop('HT115', 1)\n",
    "locations=locations.rename(columns={\"HT115\":\"HT_CCLE\"})\n",
    "locations=locations.rename(columns={\"HT115_42112\":\"HT6W56\"})\n",
    "locations=locations.rename(columns={\"HT115_42221\":\"HT6W38\"})\n",
    "locations=locations.rename(columns={\"HT115_5211\":\"HT6W44\"})\n",
    "locations=locations.rename(columns={\"HT115_7221\":\"HT6W57\"})\n",
    "col=locations.columns\n",
    "imp= filter(lambda x: 'HT' in x, col)\n",
    "print imp\n",
    "locations[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### For RPE1:\n",
    "#print locations.shape\n",
    "\n",
    "#locations=locations.rename(columns={\"RPE1_7212\":\"RPE5W44\"})\n",
    "#locations=locations.rename(columns={\"RPE1_6222\":\"RPE5W28\"})\n",
    "#locations=locations.rename(columns={\"RPE1_6221\":\"RPE5W32\"})\n",
    "#locations=locations.rename(columns={\"RPE1_621\":\"RPE5W36\"})\n",
    "#locations=locations.rename(columns={\"RPE1_42221\":\"RPE5W38\"})\n",
    "\n",
    "#col=locations.columns\n",
    "#imp= filter(lambda x: 'RPE' in x, col)\n",
    "#print imp\n",
    "#locations[0:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preapare the dataframe of the full list of candidates SNVs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#this function convert one 'spot' from the full mutation table\n",
    "def make_spot(mut,i):\n",
    "    col=mut.columns\n",
    "    imp= filter(lambda x: 'HT' in x, col)+filter(lambda x: '2nd' in x, col)\n",
    "    seq=[]\n",
    "    for f in imp:\n",
    "        seq.append(mut[f][i])\n",
    "    samples=pd.DataFrame(seq,index=imp,columns=['Seq'])\n",
    "    return samples\n",
    "\n",
    "def pre_seq_bin(seq):\n",
    "    #######################\n",
    "    #threshold for the binary is 20% \n",
    "    thr=0.2\n",
    "    binary=np.zeros(4)#'ATCG'\n",
    "    count=np.array([seq.count('A'),seq.count('T'),seq.count('C'),seq.count('G')])\n",
    "    total= len(seq)\n",
    "    percent = (count.astype(float)/total)\n",
    "    count=0\n",
    "    for i in percent:\n",
    "        if (i>thr):\n",
    "            binary[count]=1\n",
    "        count=count+1         \n",
    "    return binary\n",
    "\n",
    "#function that read seq and return matrix of counts for each letter\n",
    "def print_counts(seq):\n",
    "    count=np.array([seq.count('A'),seq.count('T'),seq.count('C'),seq.count('G')])\n",
    "    return count\n",
    "\n",
    "#function that read seq and return matrix of percentage for each letter\n",
    "def pre_seq(seq):\n",
    "    count=np.array([seq.count('A'),seq.count('T'),seq.count('C'),seq.count('G')])\n",
    "    total= len(seq)\n",
    "    percent = (count.astype(float)/total)\n",
    "    return percent\n",
    "\n",
    "#build the vector from nucleotides\n",
    "def convert_letter(char):#'ATCG'\n",
    "    if char==\"A\":\n",
    "        ans=0\n",
    "    elif char==\"T\":\n",
    "        ans=1\n",
    "    elif char==\"C\":\n",
    "        ans=2\n",
    "    elif char==\"G\":\n",
    "        ans=3\n",
    "    else:\n",
    "        ans=-1\n",
    "    return ans\n",
    "\n",
    "#show specific lines\n",
    "def present(mut,ind):\n",
    "    for i in mut.loc[ind].full_data :\n",
    "        print i[0],i[2],i[6]\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#add \"full_data\" to the mutation table:\n",
    "cwf=locations\n",
    "fmut=[]\n",
    "columns=('contig','position','context','ref_all','alt_all','t_lod_fstar','full_data')\n",
    "for i1 in cwf.index:\n",
    "    spot=make_spot(cwf,i1)\n",
    "    count=0\n",
    "    fsdata=[]\n",
    "    for i2 in spot.Seq:\n",
    "        #print (spot.index[count],\": \",print_counts(i),\" \", pre_seq_bin(i))\n",
    "        sdata=(spot.index[count],\": \",print_counts(i2),\" \",pre_seq(i2),\" \", pre_seq_bin(i2))\n",
    "        fsdata.append(sdata)\n",
    "        count=count+1\n",
    "    line=(cwf.contig[i1],cwf.position[i1],cwf.context[i1],cwf.ref_allele[i1],cwf.alt_allele[i1],cwf.t_lod_fstar[i1],fsdata)\n",
    "    fmut.append(line)\n",
    "    #cwf.set_value(index=i1,col='full_data',value=fsdata)\n",
    "full_mut = pd.DataFrame(fmut, columns=columns)\n",
    "full_mut.to_pickle('/broad/mushroom_lasagna/YB/Ht115_new_samples/full_mutations.pkl')\n",
    "print \"finish next time can skip\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
